File: backend/fathom/models/tasks.py
from typing import List, Optional
from pydantic import BaseModel


class TaskDefinitionId(BaseModel):
    scope: str
    code: str


class TaskDefinitionVersion(BaseModel):
    asAtModified: str


class TaskVersion(BaseModel):
    asAtCreated: str
    userIdCreated: str
    requestIdCreated: str
    asAtModified: str
    userIdModified: str
    requestIdModified: str
    asAtVersionNumber: int


class TaskField(BaseModel):
    name: str
    value: Optional[str] = None


class TaskReference(BaseModel):
    id: str
    taskDefinitionId: TaskDefinitionId
    taskDefinitionVersion: TaskDefinitionVersion
    taskDefinitionDisplayName: str
    state: str


class WorkflowTask(BaseModel):
    id: str
    taskDefinitionId: TaskDefinitionId
    taskDefinitionVersion: TaskDefinitionVersion
    taskDefinitionDisplayName: str
    state: str
    ultimateParentTask: Optional[TaskReference] = None
    parentTask: Optional[TaskReference] = None
    childTasks: List[TaskReference] = []
    correlationIds: List[str] = []
    version: TaskVersion
    terminalState: bool
    asAtLastTransition: str
    fields: List[TaskField] = []
    stackingKey: Optional[str] = None
    actionLogIdCreated: Optional[str] = None
    actionLogIdModified: Optional[str] = None
    actionLogIdSubmitted: Optional[str] = None


class TaskListResponse(BaseModel):
    values: List[WorkflowTask]
    href: str
    links: List[dict] = []
    nextPage: Optional[str] = None
    previousPage: Optional[str] = None


class TaskFilter(BaseModel):
    dateFrom: Optional[str] = None
    dateTo: Optional[str] = None
    searchQuery: Optional[str] = None
    states: Optional[List[str]] = None
    correlationIds: Optional[List[str]] = None

File: agent-ui/src/types/tasks.ts
// LUSID Workflow Task Types - Based on real API responses

export interface TaskDefinitionId {
  scope: string
  code: string
}

export interface TaskDefinitionVersion {
  asAtModified: string
}

export interface TaskVersion {
  asAtCreated: string
  userIdCreated: string
  requestIdCreated: string
  asAtModified: string
  userIdModified: string
  requestIdModified: string
  asAtVersionNumber: number
}

export interface TaskField {
  name: string
  value: string | number | null
}

export interface TaskReference {
  id: string
  taskDefinitionId: TaskDefinitionId
  taskDefinitionVersion: TaskDefinitionVersion
  taskDefinitionDisplayName: string
  state: string
}

export interface WorkflowTask {
  id: string
  taskDefinitionId: TaskDefinitionId
  taskDefinitionVersion: TaskDefinitionVersion
  taskDefinitionDisplayName: string
  state: TaskState
  ultimateParentTask?: TaskReference
  parentTask?: TaskReference
  childTasks: TaskReference[]
  correlationIds: string[]
  version: TaskVersion
  terminalState: boolean
  asAtLastTransition: string
  fields: TaskField[]
  stackingKey?: string
  actionLogIdCreated?: string
  actionLogIdModified?: string
  actionLogIdSubmitted?: string
}

// Aggregation Error specific states
export type UltimateParentState = 'Pending' | 'Searching Errors' | 'inReview' | 'Completed'
export type ChildTaskState = 'Pending' | 'InReview' | 'Resolved'
export type TaskState = UltimateParentState | ChildTaskState

export interface TaskListResponse {
  nextPage?: string
  previousPage?: string
  values: WorkflowTask[]
  href: string
  links: Array<{
    relation: string
    href: string
    description: string
    method: string
  }>
}

// UI-specific types
export interface TaskGroup {
  ultimateParent: WorkflowTask
  children: WorkflowTask[]
  totalCount: number
}

export interface TaskFilter {
  dateFrom?: string
  dateTo?: string
  searchQuery?: string
  states?: TaskState[]
  correlationIds?: string[]
}

// Store interface extension
export interface TaskStore {
  selectedTasks: WorkflowTask[]
  taskGroups: TaskGroup[]
  taskFilter: TaskFilter
  isTaskPanelVisible: boolean
  
  // Actions
  addSelectedTask: (task: WorkflowTask) => void
  removeSelectedTask: (taskId: string) => void
  addSelectedTaskGroup: (group: TaskGroup) => void
  removeSelectedTaskGroup: (ultimateParentId: string) => void
  setTaskFilter: (filter: Partial<TaskFilter>) => void
  setTaskGroups: (groups: TaskGroup[]) => void
  clearSelectedTasks: () => void
  setTaskPanelVisible: (visible: boolean) => void
}
